server/task_executor_pitr_cutover.go

```diff
@@ -137,3 +145,115 @@ func (exec *PITRCutoverTaskExecutor) pitrCutover(ctx context.Context, task *api.
 		Detail: fmt.Sprintf("Swapped PITR database for target database %q", task.Database.Name),
 	}, nil
 }
+
+func swapDatabasesAndPrepareForBackup(ctx context.Context, conn *sql.Conn, database *api.Database, suffixTs int64) (*sql.Tx, string, error) {
+	log.Debug("Acquiring table locks in database", zap.String("database", database.Name))
+	if err := mysql.FlushTablesWithReadLock(ctx, conn, database.Name); err != nil {
+		return nil, "", err
+	}
+	log.Debug("Swapping the original and PITR database", zap.String("originalDatabase", database.Name))
+	pitrDatabaseName, pitrOldDatabaseName, err := mysql.SwapPITRDatabase(ctx, conn, database.Name, suffixTs)
+	if err != nil {
+		log.Error("Failed to swap the original and PITR database", zap.String("originalDatabase", database.Name), zap.String("pitrDatabase", pitrDatabaseName), zap.Error(err))
+		return nil, "", fmt.Errorf("failed to swap the original and PITR database, error: %w", err)
+	}
+	log.Debug("Finished swapping the original and PITR database", zap.String("originalDatabase", database.Name), zap.String("pitrDatabase", pitrDatabaseName), zap.String("oldDatabase", pitrOldDatabaseName))
+
+	binlogInfo, err := mysql.GetBinlogInfo(ctx, conn)
+	if err != nil {
+		return nil, "", err
+	}
+	payload := api.BackupPayload{BinlogInfo: binlogInfo}
+	payloadBytes, err := json.Marshal(payload)
+	if err != nil {
+		return nil, "", err
+	}
+
+	log.Debug("Starting new transaction and unlock tables")
+	options := sql.TxOptions{ReadOnly: true}
+	// Beginning a transaction in the same session will implicitly release existing table locks.
+	// ref: https://dev.mysql.com/doc/refman/8.0/en/lock-tables.html, section "Interaction of Table Locking and Transactions".
+	txn, err := conn.BeginTx(ctx, &options)
+	if err != nil {
+		return nil, "", err
+	}
+
+	return txn, string(payloadBytes), nil
+}
+
+func (exec *PITRCutoverTaskExecutor) backupDatabaseAfterPITR(ctx context.Context, conn *sql.Conn, txn *sql.Tx, backupPayload string, driver db.Driver, store *store.Store, database *api.Database, dataDir string, suffixTs int64) error {
+	backupName := fmt.Sprintf("%s-%s-pitr_%d", api.ProjectShortSlug(database.Project), api.EnvSlug(database.Instance.Environment), suffixTs)
+	if err := createBackupDirectory(dataDir, database.ID); err != nil {
+		return fmt.Errorf("failed to create backup directory, error: %w", err)
+	}
+	path := getBackupRelativeFilePath(database.ID, backupName)
```

Comment:
Also, refactor this to a common function.

We can create the backup asynchronously similar to the other two usages.

Reply:
We cannot schedule the backup task to a pipeline because we need to dump the database using the very transaction we created here. The locking and binlog acquiring process is slightly different from a normal backup process.